File: BazaMobileHW/src/main/NotesApp/src/Menu.kt
import java.util.UUID

class Menu(
    private val console: ConsoleUI,
    private val notebook: Notebook
) {
    private enum class State {
        HOME,
        TEXT_NOTE,
        REMINDER_NOTE,
        TEXT_NEW_NOTE,
        REMINDER_NEW_NOTE
    }

    private var currentState: State = State.HOME
    private var currentNote: NoteProtocol? = null

    fun run() {
        var shouldExit = false

        while (!shouldExit) {
            when (currentState) {
                State.HOME -> shouldExit = showHomeMenu()
                State.TEXT_NOTE -> showNoteMenu()
                State.REMINDER_NOTE -> showNoteMenu()
                State.TEXT_NEW_NOTE -> createNewTextNote()
                State.REMINDER_NEW_NOTE -> createNewReminderNote()
            }
        }
        println("До свидания!")
    }

    private fun showHomeMenu(): Boolean {
        console.showCanvas(notebook)

        val menuOptions = listOf(
            "Редактировать заметку",
            "Добавить новую заметку",
            "Удалить заметку",
            "Выйти из приложения"
        )

        return when (console.showMenuList(menuOptions)) {
            1 -> {
                selectNoteIndex()?.let { editNoteAt(it) }
                false
            }
            2 -> {
                showAddNoteMenu()
                false
            }
            3 -> {
                selectNoteIndex()?.let { removeNoteAt(it) }
                false
            }
            4 -> true
            else -> {
                println("Неверный ввод")
                false
            }
        }
    }

    private fun selectNoteIndex(): Int? {
        val notes = notebook.getAllNotes()
        if (notes.isEmpty()) {
            println("Заметок нет")
            return null
        }

        val index = console.readInt("Выберите номер заметки (1-${notes.size})") ?: return null
        val adjustedIndex = index - 1

        return if (adjustedIndex in notes.indices) adjustedIndex else {
            println("Неверный номер заметки")
            null
        }
    }

    private fun showAddNoteMenu() {
        val addOptions = listOf(
            "Текстовая заметка",
            "Заметка-напоминание"
        )

        when (console.showMenuList(addOptions)) {
            1 -> currentState = State.TEXT_NEW_NOTE
            2 -> currentState = State.REMINDER_NEW_NOTE
            else -> println("Неверный ввод")
        }
    }

    private fun createNewTextNote() {
        println("\nСоздание текстовой заметки")

        val name = console.readString("Введите название:") ?: run {
            println("Отмена создания заметки")
            currentState = State.HOME
            return
        }

        val text = console.readString("Введите текст:") ?: run {
            println("Отмена создания заметки")
            currentState = State.HOME
            return
        }

        val textNote = notebook.createTextNote(name, text)
        notebook.add(textNote)
        println("Текстовая заметка создана!")
        currentState = State.HOME
    }

    private fun createNewReminderNote() {
        println("\nСоздание заметки-напоминания")

        val text = console.readString("Введите текст напоминания:") ?: run {
            println("Отмена создания заметки")
            currentState = State.HOME
            return
        }

        val reminderNote = notebook.createReminderNote(text)
        notebook.add(reminderNote)
        println("Заметка-напоминание создана!")
        currentState = State.HOME
    }

    private fun editNoteAt(index: Int) {
        val note = notebook.getAllNotes()[index]
        currentNote = note
        currentState = when (note) {
            is TextNote -> State.TEXT_NOTE
            is ReminderNote -> State.REMINDER_NOTE
            else -> State.HOME
        }
    }

    private fun showNoteMenu() {
        currentNote?.let { console.showCanvas(it) }

        val options = when (currentNote) {
            is TextNote -> listOf("Изменить название", "Изменить текст", "Назад")
            is ReminderNote -> listOf("Изменить текст", "Переключить статус", "Назад")
            else -> listOf("Назад")
        }

        when (console.showMenuList(options)) {
            1 -> when (currentNote) {
                is TextNote -> updateTextNoteName()
                is ReminderNote -> updateReminderText()
                else -> Unit
            }
            2 -> when (currentNote) {
                is TextNote -> updateTextNoteText()
                is ReminderNote -> toggleReminderStatus()
                else -> Unit
            }
            else -> {
                currentState = State.HOME
                currentNote = null
            }
        }
    }

    private fun updateTextNoteName() {
        (currentNote as? TextNote)?.let { note ->
            console.readString("Новое название:")?.let { newName ->
                val newData = note.data.copy(name = newName)
                note.update(newData)
                println("Название обновлено")
            }
        }
    }

    private fun updateTextNoteText() {
        (currentNote as? TextNote)?.let { note ->
            console.readString("Новый текст:")?.let { newText ->
                val newData = note.data.copy(text = newText)
                note.update(newData)
                println("Текст обновлен")
            }
        }
    }

    private fun updateReminderText() {
        (currentNote as? ReminderNote)?.let { note ->
            console.readString("Новый текст:")?.let { newText ->
                val newData = note.data.copy(text = newText)
                note.update(newData)
                println("Текст обновлен")
            }
        }
    }

    private fun toggleReminderStatus() {
        (currentNote as? ReminderNote)?.let { note ->
            val newData = note.data.copy(isDone = !note.data.isDone)
            note.update(newData)
            println("Статус обновлен: ${if (newData.isDone) "Выполнено" else "Не выполнено"}")
        }
    }

    private fun removeNoteAt(index: Int) {
        notebook.removeAt(index)
        println("Заметка удалена!")
    }
}

File: NotesApp/menu.swift
import Foundation
class Menu {
    private let console: ConsoleUI
    private let notebook: Notebook
    
    private enum State {
        case home
        case textNote(TextNote)
        case reminderNote(ReminderNote)
        case textNewNote
        case reminderNewNote
    }
    
    private var currentState: State = .home
    
    init(console: ConsoleUI, notebook: Notebook) {
        self.console = console
        self.notebook = notebook
    }
    
    func run() {
        var shouldExit = false
        
        while !shouldExit {
            switch currentState {
            case .home:
                shouldExit = showHomeMenu()
            case .textNote(let note):
                showNoteMenu(note: note)
            case .reminderNote(let note):
                showNoteMenu(note: note)
            case .textNewNote:
                createNewTextNote()
            case .reminderNewNote:
                createNewReminderNote()
            }
        }
        print("До свидания!")
    }
    
    private func showHomeMenu() -> Bool {
        console.showCanvas(notebook)
        
        let menuOptions = [
            "Редактировать заметку",
            "Добавить новую заметку",
            "Удалить заметку",
            "Выйти из приложения"
        ]
        
        guard let choice = console.showMenuList(menuOptions) else {
            print(" Неверный ввод")
            return false
        }
        
        switch choice {
        case 1:
            if let index = selectNoteIndex() {
                editNote(at: index)
            }
            return false
            
        case 2:
            showAddNoteMenu()
            return false
            
        case 3:
            if let index = selectNoteIndex() {
                removeNote(at: index)
            }
            return false
            
        case 4:
            return true
            
        default:
            print(" Неверный выбор")
            return false
        }
    }
    
    private func selectNoteIndex() -> Int? {
        let notes = notebook.getAllNotes()
        if notes.isEmpty {
            print("📭 Заметок нет")
            return nil
        }
        
        guard let index = console.readInt("Выберите номер заметки (1-\(notes.count)): ") else {
            print("Неверный ввод")
            return nil
        }
        
        let adjustedIndex = index - 1
        guard adjustedIndex >= 0 && adjustedIndex < notes.count else {
            print("Неверный номер заметки")
            return nil
        }
        
        return adjustedIndex
    }
    
    private func showAddNoteMenu() {
        let addOptions = [
            "Текстовая заметка",
            "Заметка-напоминание"
        ]
        
        guard let choice = console.showMenuList(addOptions) else {
            print(" Неверный ввод")
            return
        }
        
        switch choice {
        case 1:
            currentState = .textNewNote
        case 2:
            currentState = .reminderNewNote
        default:
            print(" Неверный выбор")
        }
    }
    
    private func createNewTextNote() {
        print("\n Создание текстовой заметки")
        
        guard let name = console.readString("Введите название: "),
              let text = console.readString("Введите текст: ") else {
            print(" Отмена создания заметки")
            currentState = .home
            return
        }
        
        let id = UUID().uuidString
        let model = TextNoteModel(id: id, name: name, text: text)
        
        let textNote = TextNote(data: model, storage: notebook.textStorage)
        notebook.add(textNote)
        
        print(" Текстовая заметка создана!")
        currentState = .home
    }
    
    private func createNewReminderNote() {
        print("\n Создание заметки-напоминания")
        
        guard let text = console.readString("Введите текст напоминания: ") else {
            print(" Отмена создания заметки")
            currentState = .home
            return
        }
        
        let id = UUID().uuidString
        let model = ReminderNoteModel(id: id, text: text, isDone: false)
        
        let reminderNote = ReminderNote(data: model, storage: Storage<ReminderNoteModel>())
        notebook.add(reminderNote)
        
        print(" Заметка-напоминание создана!")
        currentState = .home
    }
    
    private func editNote(at index: Int) {
        let notes = notebook.getAllNotes()
        let note = notes[index]
        
        if let textNote = note as? TextNote {
            currentState = .textNote(textNote)
        } else if let reminderNote = note as? ReminderNote {
            currentState = .reminderNote(reminderNote)
        }
    }
    
    private func showNoteMenu(note: NoteProtocol) {
        console.showCanvas(note)
        
        var options: [String]
        var isReminder = false
        
        if note is TextNote {
            options = ["Изменить название", "Изменить текст", "Назад"]
        } else if let reminderNote = note as? ReminderNote {
            options = ["Изменить текст", "Переключить статус", "Назад"]
            isReminder = true
        } else {
            options = ["Назад"]
        }
        
        guard let choice = console.showMenuList(options) else {
            currentState = .home
            return
        }
        
        switch (isReminder, choice) {
        case (false, 1):
            if let textNote = note as? TextNote,
               let newName = console.readString("Новое название: ") {
                var model = textNote.data
                model.name = newName
                textNote.update(model)
            }
            
        case (false, 2):
            if let textNote = note as? TextNote,
               let newText = console.readString("Новый текст: ") {
                var model = textNote.data
                model.text = newText
                textNote.update(model)
            }
            
        case (true, 1):
            if let reminderNote = note as? ReminderNote,
               let newText = console.readString("Новычй текст: ") {
                var model = reminderNote.data
                model.text = newText
                reminderNote.update(model)
            }
            
        case (true, 2):
            if let reminderNote = note as? ReminderNote {
                var model = reminderNote.data
                model.isDone.toggle()
                reminderNote.update(model)
                print(" Статус обновлен: \(model.isDone ? " Выполнено" : " Не выполнено")")
            }
            
        default:
            currentState = .home
        }
    }
    
    private func removeNote(at index: Int) {
        let notes = notebook.getAllNotes()
        let note = notes[index]
        
        note.willRemove()
        notebook.remove(at: index)
        
        print(" Заметка удалена!")
    }
}
